---
title: "N2_L4_RNAseq (intestine)"
author: "Jessica Hill"
date: "`r Sys.Date()`"
output: html_document
---

all whole worm controls have been taken out!
that will be a separate analysis just for evaluating purity


http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

install packages
```{r}
# If you don't have bioconductor, install it. This version works with R version 4.1. Use "3.11" for R version 4.0

#if (!require("BiocManager", quietly = TRUE))
#        install.packages("BiocManager")
#BiocManager::install(version = "3.16")

#if (!requireNamespace("BiocManager", quietly = TRUE))
#        install.packages("BiocManager")

#BiocManager::install("DESeq2")
#BiocManager::install("apeglm")
#install.packages("corrplot")
#install.packages("pheatmap")
#install.packages("RColorBrewer")
#BiocManager::install("topGO")

#install.packages("calibrate")
#install.packages('ashr')
```

load packages
```{r load-packages, include=FALSE}
library(DESeq2)
library(corrplot)
library(RColorBrewer)
library(pheatmap)
library(apeglm)
library(dplyr)
library(tidyverse)
library(biomaRt)
library(calibrate)
library(ashr)
```

read in the data
```{r}
data <- read.table(file = "01_input/all.counts.txt", header = TRUE)
```

Prepare our input data for analysis.
We read in a count matrix, which we will name counts, and the sample information table, which we will name coldata.
Generate a table called "counts" out of the countsData table.
```{r}
colnames(data)

counts <- as.matrix(data[,6:28]) # we should remove Rep3 OP50 whole worm because it is a shitty sample

counts_int <- counts[, grep("intestine", colnames(counts))]
str(counts_int)
```


```{r}
# what are the original column names?
colnames(counts_int)

# now reassign 
colnames(counts_int) <- c('Rep1.CeMbio.intestine','Rep1.CeMbio.to.PA14.intestine','Rep1.OP50.intestine','Rep1.OP50.to.PA14.intestine','Rep2.CeMbio.intestine', 'Rep2.CeMbio.to.PA14.intestine','Rep2.OP50.intestine', 'Rep2.OP50.to.PA14.intestine','Rep3.CeMbio.intestine','Rep3.CeMbio.to.PA14.intestine','Rep3.OP50.intestine','Rep3.OP50.to.PA14.intestine')

dim(counts_int)
```

Now make the sample information table, which we will name coldata.
```{r}
diet <- c('CeMbio','CeMbio','OP50','OP50','CeMbio','CeMbio','OP50','OP50','CeMbio','CeMbio','OP50','OP50')
exposure <- c('no','yes','no','yes','no','yes','no','yes','no','yes','no','yes')
rep <- c('rep1','rep1','rep1','rep1','rep2', 'rep2', 'rep2', 'rep2','rep3', 'rep3', 'rep3', 'rep3')
group <- c('CeMbio_no','CeMbio_yes','OP50_no','OP50_yes','CeMbio_no','CeMbio_yes','OP50_no','OP50_yes','CeMbio_no','CeMbio_yes','OP50_no','OP50_yes')
sampleID <- c('Rep1.CeMbio.intestine','Rep1.CeMbio.to.PA14.intestine','Rep1.OP50.intestine', 'Rep1.OP50.to.PA14.intestine','Rep2.CeMbio.intestine', 'Rep2.CeMbio.to.PA14.intestine', 'Rep2.OP50.intestine', 'Rep2.OP50.to.PA14.intestine','Rep3.CeMbio.intestine','Rep3.CeMbio.to.PA14.intestine','Rep3.OP50.intestine', 'Rep3.OP50.to.PA14.intestine')
  
coldata <- data.frame(sampleID, diet, exposure, group, rep)
coldata

rownames(coldata)<- coldata$sampleID
coldata
```


Create an ddsHTSeq object out of counts and coldata.
This will set a base design for your experiment.
Load the counts data and to attach it/them to the metadata.
```{r}
# design 1 - (genotype + condition) the exposure effect represents the overall effect controlling for differences due to diet
dds <- DESeqDataSetFromMatrix(countData = counts_int,
                              colData = coldata,
                              design = ~ group)

# design 2 - (genotype:condition) the main condition effect only represents the effect of condition for the reference level of genotype (I, or whichever level was defined by the user as the reference level).
#dds1 <- DESeqDataSetFromMatrix(countData = counts_int,
#                              colData = coldata,
#                              design = ~ diet + exposure + diet:exposure)


# how many genes do we have?
dim(dds) #20447
```


pre-filtering (filter for present genes). 
Not necessary, but helps keep things fast. 
```{r}
#Exclude all samples that have less than 10 reads:
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
#How many are we left with?
dim(dds) # 18265
```

Visualize read count distribution
```{r}
raw_count_threshold <- 10
hist(log10(rowSums(counts(dds))), breaks = 50)
abline(v = log10(raw_count_threshold), col = "red", lty = 2)
```

Factor Levels
Organize the categories into an order that makes sense
```{r}
dds$diet <- factor(dds$diet, levels = c("OP50", "CeMbio")) # this sets OP50 as the reference levels
dds$exposure <- factor(dds$exposure, levels = c("no", "yes")) # this sets no as the reference levels
```


differential expression analysis 
Perform DESEQ2 analysis
```{r}
#Will transform dds into a specialized object
dds <- DESeq(dds)
dim(dds)
plotDispEsts(dds)

# Here is a demonstration of the size Factor scaling that was calculated (sizeFactor):
dds$sizeFactor

#Access the normalized counts using counts(x, normalized = TRUE)
#Access the raw count info using counts(x, normalized = FALSE)
head(counts(dds, normalized = TRUE))
#head(counts(dds, normalized = FALSE))
```


extract results 
```{r}
#calculate the statistically significant differences between CeMbio and OP50 diets
res_CeMbio.yes_CeMbio.no <- results(dds, 
                                    lfc = 0.5,
                                    contrast = c("group", "CeMbio_yes", "CeMbio_no"))
summary(res_CeMbio.yes_CeMbio.no)

res_CeMbio.yes_OP50.yes <- results(dds, 
                                    lfc = 0.5,
                                    contrast = c("group", "CeMbio_yes", "OP50_yes"))
summary(res_CeMbio.yes_OP50.yes)

res_CeMbio.no_OP50.no <- results(dds, 
                                    lfc = 0.5,
                                    contrast = c("group", "CeMbio_no", "OP50_no"))
summary(res_CeMbio.no_OP50.no)

res_OP50.yes_OP50.no <- results(dds, 
                                    lfc = 0.5,
                                    contrast = c("group", "OP50_yes", "OP50_no"))
summary(res_OP50.yes_OP50.no)
```

```{r}
#res_diet1 <- results(dds1,
#                    lfc = 0.5,
#                    contrast=c("diet", "CeMbio", "OP50")) # the diet effect for no exposure
#summary(res_diet1)

#res_exposure1 <- results(dds1,
#                    lfc = 0.5,
#                    contrast=c("exposure", "yes", "no")) # the exposure effect for OP50
#summary(res_exposure1)

#res_interaction1 <- results(dds1,
#                            lfc = 0.5,
#                            contrast=list(c("exposure_yes_vs_no", "dietCeMbio.exposureyes"))) # the exposure effect for CeMbio
#summary(res_interaction1)
```

```{r}
resultsNames(dds)
```

perform log fold change shrinkage for visualization
```{r}
reslfc_CeMbio.yes_CeMbio.no <- lfcShrink(dds, contrast = c("group", "CeMbio_yes", "CeMbio_no"), type="ashr")

reslfc_CeMbio.yes_OP50.yes <- lfcShrink(dds, contrast = c("group", "CeMbio_yes", "OP50_yes"), type="ashr")  

reslfc_CeMbio.no_OP50.no <- lfcShrink(dds, contrast = c("group", "CeMbio_no", "OP50_no"), type="ashr")

reslfc_OP50.yes_OP50.no <- lfcShrink(dds, contrast = c("group", "OP50_yes", "OP50_no"), type="ashr")
```

```{r}
#reslfc_diet1 <- lfcShrink(dds2, contrast = c("diet", "CeMbio", "OP50"), type="ashr")
  
#reslfc_exposure1 <- lfcShrink(dds2, contrast = c("exposure", "yes", "no"), type="ashr")  

#reslfc_interaction1 <- lfcShrink(dds2, contrast = list(c("exposure_yes_vs_no", "dietCeMbio.exposureyes")), type="ashr")
```


Exploring and exporting results
MA plot
```{r}
# save plot as a .png
png("03_output/intestine/MA plot - CeMbio.yes.vs.CeMbio.no.png")
# Reset parameters
# Reset plotting parameters to their default values
par(mfrow=c(1,1))
# create the plot
plotMA(reslfc_CeMbio.yes_CeMbio.no, ylim = c(-7,7), main = "Effect of exposure to PA14 on CeMbio fed worms", 
       ylab = "log fold change" # (ratio of normalized CeMbio.yes / CeMbio.no)
       , xlab = "means of normalized counts")
# Increase the font size of the axis labels
par(cex.lab=2)

dev.off()
```
```{r}
# save plot as a .png
png("03_output/intestine/MA plot - CeMbio.yes.vs.OP50.yes.png")
# Reset parameters
# Reset plotting parameters to their default values
par(mfrow=c(1,1))
# create the plot
plotMA(reslfc_CeMbio.yes_OP50.yes, ylim = c(-7,7), main = "Effect of diet during exposure to PA14", 
       ylab = "log fold change" # (ratio of normalized CeMbio.yes / OP50.yes)
       , xlab = "means of normalized counts")
# Increase the font size of the axis labels
par(cex.lab=2)

dev.off()
```
```{r}
# save plot as a .png
png("03_output/intestine/MA plot - CeMbio.no.vs.OP50.no.png")
# Reset parameters
# Reset plotting parameters to their default values
par(mfrow=c(1,1))
# create the plot
plotMA(reslfc_CeMbio.no_OP50.no, ylim = c(-7,7), main = "Effect of diet (CeMbio vs OP50)", 
       ylab = "log fold change" # (ratio of normalized CeMbio.no / OP50.no)
       , xlab = "means of normalized counts")
# Increase the font size of the axis labels
par(cex.lab=2)

dev.off()
```
```{r}
# save plot as a .png
png("03_output/intestine/MA plot - OP50.yes.vs.OP50.no.png")
# Reset parameters
# Reset plotting parameters to their default values
par(mfrow=c(1,1))
# create the plot
plotMA(reslfc_OP50.yes_OP50.no, ylim = c(-7,7), main = "Effect of exposure to PA14 on OP50 fed worms", 
       ylab = "log fold change" # (ratio of normalized OP50.yes / OP50.no)
       , xlab = "means of normalized counts")
# Increase the font size of the axis labels
par(cex.lab=2)

dev.off()
```


Sample distance matrix
```{r}
rld <- rlog(dds, blind=FALSE)
new.rld <- assay(rld)

rld.CeMbio <- new.rld[, grep("CeMbio", colnames(new.rld))]
rld.OP50 <- new.rld[, grep("OP50", colnames(new.rld))]

# Calculate the distances between each sample
#sampleDists <- dist(t(assay(rld))) # original code, not using
sampleDists <- dist(t(new.rld))
sampleDists.CeMbio <- dist(t(rld.CeMbio))
sampleDists.OP50 <- dist(t(rld.OP50))

# Convert from data.frame to matrix
sampleDistMatrix <- as.matrix(sampleDists) 
sampleDistMatrix.CeMbio <- as.matrix(sampleDists.CeMbio)
sampleDistMatrix.OP50 <- as.matrix(sampleDists.OP50)

# Add some labels
rownames(sampleDistMatrix) <- colnames(new.rld)
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette(rev(brewer.pal(9, "Blues")))(255)

rownames(sampleDistMatrix.CeMbio) <- colnames(rld.CeMbio)
colnames(sampleDistMatrix.CeMbio) <- NULL
colors <- colorRampPalette(rev(brewer.pal(9, "Blues")))(255)

rownames(sampleDistMatrix.OP50) <- colnames(rld.OP50)
colnames(sampleDistMatrix.OP50) <- NULL
colors <- colorRampPalette(rev(brewer.pal(9, "Blues")))(255)


# Draw the heatmap
png("03_output/intestine/HeatMap - all.png")
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors, fontsize = 10, treeheight_col = 20, treeheight_row = 20,
         legend = FALSE)

# save plot as a .png
png("03_output/intestine/HeatMap - CeMbio.png")
pheatmap(sampleDistMatrix.CeMbio,
         clustering_distance_rows=sampleDists.CeMbio,
         clustering_distance_cols=sampleDists.CeMbio,
         col=colors, fontsize = 10, treeheight_col = 20, treeheight_row = 20,
         legend = FALSE)

png("03_output/intestine/HeatMap - OP50.png")
pheatmap(sampleDistMatrix.OP50,
         clustering_distance_rows=sampleDists.OP50,
         clustering_distance_cols=sampleDists.OP50,
         col=colors, fontsize = 10, treeheight_col = 20, treeheight_row = 20,
         legend = FALSE)

dev.off()
```


Volcano plots
Display the fold change against the p-value
```{r}
#Identify significantly changing genes
significantLFC.1 <- rbind(subset(reslfc_CeMbio.yes_CeMbio.no, padj < 0.1 & log2FoldChange > 0.5), subset(reslfc_CeMbio.yes_CeMbio.no, padj < 0.1 & log2FoldChange < -0.5))

significantLFC.2 <- rbind(subset(reslfc_CeMbio.yes_OP50.yes, padj < 0.1 & log2FoldChange > 0.5), subset(reslfc_CeMbio.yes_OP50.yes, padj < 0.1 & log2FoldChange < -0.5))

significantLFC.3 <- rbind(subset(reslfc_CeMbio.no_OP50.no, padj < 0.1 & log2FoldChange > 0.5), subset(reslfc_CeMbio.no_OP50.no, padj < 0.1 & log2FoldChange < -0.5))

significantLFC.4 <- rbind(subset(reslfc_OP50.yes_OP50.no, padj < 0.1 & log2FoldChange > 0.5), subset(reslfc_OP50.yes_OP50.no, padj < 0.1 & log2FoldChange < -0.5))

############################
significant_points_to_plot.1 <-reslfc_CeMbio.yes_CeMbio.no[which(rownames(reslfc_CeMbio.yes_CeMbio.no) %in% rownames(significantLFC.1)),] 

significant_points_to_plot.2 <-reslfc_CeMbio.yes_OP50.yes[which(rownames(reslfc_CeMbio.yes_OP50.yes) %in% rownames(significantLFC.2)),]

significant_points_to_plot.3 <-reslfc_CeMbio.no_OP50.no[which(rownames(reslfc_CeMbio.no_OP50.no) %in% rownames(significantLFC.3)),]

significant_points_to_plot.4 <-reslfc_OP50.yes_OP50.no[which(rownames(reslfc_OP50.yes_OP50.no) %in% rownames(significantLFC.4)),]


# We will set the top limit of the plot as 100. Need to find all the points that exceed that measure
maxedout.1 <- subset(reslfc_CeMbio.yes_CeMbio.no, padj < 10e-100)
maxedout.2 <- subset(reslfc_CeMbio.yes_OP50.yes, padj < 10e-100)
maxedout.3 <- subset(reslfc_CeMbio.no_OP50.no, padj < 10e-100)
maxedout.4 <- subset(reslfc_OP50.yes_OP50.no, padj < 10e-100)


png("03_output/intestine/Volcano plot - CeMbio.yes_CeMbio.no.png")
plot(reslfc_CeMbio.yes_CeMbio.no$log2FoldChange, -log10(reslfc_CeMbio.yes_CeMbio.no$padj),
     main="Effect of exposure to PA14 on CeMbio fed worms", xlab="Effect size: log2(fold-change)", ylab="-log10(adjusted p-value)", pch=20, cex = 0.75, ylim = c(0, 20), xlim = c(-15,15), col = "#00000030")
# Add points
points(maxedout.1$log2FoldChange, rep(102, length(maxedout.1$log2FoldChange)),
       pch=20, cex = 0.75, ylim = c(0, 100), col = "red")
points(significant_points_to_plot.1$log2FoldChange, -log10(significant_points_to_plot.1$padj),
       pch=20, cex = 0.75, ylim = c(0, 100), col = "red")
# Add lines
abline(v=0, col = "blue")
abline(v=0.5, col = "blue", lty = "dashed")
abline(v=-0.5, col = "blue", lty = "dashed")
abline(h=-log10(0.1), col = "blue", lty = "dashed")

dev.off()
```
```{r}
png("03_output/intestine/Volcano plot - CeMbio.yes_OP50.yes.png")
plot(reslfc_CeMbio.yes_OP50.yes$log2FoldChange, -log10(reslfc_CeMbio.yes_OP50.yes$padj),
     main="Effect of diet during exposure to PA14", xlab="Effect size: log2(fold-change)", ylab="-log10(adjusted p-value)", pch=20, cex = 0.75, ylim = c(0, 20), xlim = c(-15,15), col = "#00000030")
# Add points
points(maxedout.2$log2FoldChange, rep(102, length(maxedout.2$log2FoldChange)),
       pch=20, cex = 0.75, ylim = c(0, 100), col = "red")
points(significant_points_to_plot.2$log2FoldChange, -log10(significant_points_to_plot.2$padj),
       pch=20, cex = 0.75, ylim = c(0, 100), col = "red")
# Add lines
abline(v=0, col = "blue")
abline(v=0.5, col = "blue", lty = "dashed")
abline(v=-0.5, col = "blue", lty = "dashed")
abline(h=-log10(0.1), col = "blue", lty = "dashed")

dev.off()
```
```{r}
png("03_output/intestine/Volcano plot - CeMbio.no_OP50.no.png")
plot(reslfc_CeMbio.no_OP50.no$log2FoldChange, -log10(reslfc_CeMbio.no_OP50.no$padj),
     main="Effect of diet (CeMbio vs OP50)", xlab="Effect size: log2(fold-change)", ylab="-log10(adjusted p-value)", pch=20, cex = 0.75, ylim = c(0, 20), xlim = c(-15,15), col = "#00000030")
# Add points
points(maxedout.3$log2FoldChange, rep(102, length(maxedout.3$log2FoldChange)),
       pch=20, cex = 0.75, ylim = c(0, 100), col = "red")
points(significant_points_to_plot.3$log2FoldChange, -log10(significant_points_to_plot.3$padj),
       pch=20, cex = 0.75, ylim = c(0, 100), col = "red")
# Add lines
abline(v=0, col = "blue")
abline(v=0.5, col = "blue", lty = "dashed")
abline(v=-0.5, col = "blue", lty = "dashed")
abline(h=-log10(0.1), col = "blue", lty = "dashed")

dev.off()
```
```{r}
png("03_output/intestine/Volcano plot - OP50.yes_OP50.no.png")
plot(reslfc_OP50.yes_OP50.no$log2FoldChange, -log10(reslfc_OP50.yes_OP50.no$padj),
     main="Effect of exposure to PA14 on OP50 fed worms", xlab="Effect size: log2(fold-change)", ylab="-log10(adjusted p-value)", pch=20, cex = 0.75, ylim = c(0, 20), xlim = c(-15,15), col = "#00000030")
# Add points
points(maxedout.4$log2FoldChange, rep(102, length(maxedout.4$log2FoldChange)),
       pch=20, cex = 0.75, ylim = c(0, 100), col = "red")
points(significant_points_to_plot.4$log2FoldChange, -log10(significant_points_to_plot.4$padj),
       pch=20, cex = 0.75, ylim = c(0, 100), col = "red")
# Add lines
abline(v=0, col = "blue")
abline(v=0.5, col = "blue", lty = "dashed")
abline(v=-0.5, col = "blue", lty = "dashed")
abline(h=-log10(0.1), col = "blue", lty = "dashed")

dev.off()
```


```{r}
png("03_output/intestine/PCA.png")
plotPCA(rld, intgroup=c("diet", "exposure", "rep"))

dev.off()
```


Check contamination
```{r}
## positive intestine controls
# elt-2 = WBGene00001250
# elt-7 = WBGene00015981
# ges-1 = WBGene00001578

## germline
# glh-1 = WBGene00001598
# nos-2 = WBGene00003784 

## muscle
# myo-3 = WBGene00003515

## neuron
# hlh-2 = WBGene00001949
# rab-3 = WBGene00004267

reslfc_CeMbio.yes_CeMbio.no
reslfc_CeMbio.yes_OP50.yes
reslfc_CeMbio.no_OP50.no
reslfc_OP50.yes_OP50.no 

plotCounts(dds, gene=which(rownames(reslfc_CeMbio.yes_CeMbio.no) == "WBGene00003515"), intgroup="group")
```


Get list of significantly changing genes
We will use the set of significantly changing genes from our variance shrunken analysis
```{r}
# reference levels = diet: OP50, and exposure: no
# Select the significant subset of genes that are up-regulated and down-regulated
reslfc_CeMbio.yes_CeMbio.no
reslfc_CeMbio.yes_OP50.yes
reslfc_CeMbio.no_OP50.no
reslfc_OP50.yes_OP50.no 


Up_CeMbio.yes_CeMbio.no <- subset(reslfc_CeMbio.yes_CeMbio.no, padj < 0.1 & log2FoldChange > 0.5)
Up_CeMbio.yes_CeMbio.no <- Up_CeMbio.yes_CeMbio.no[order(Up_CeMbio.yes_CeMbio.no$padj),]

Down_CeMbio.yes_CeMbio.no <- subset(reslfc_CeMbio.yes_CeMbio.no, padj < 0.1 & log2FoldChange > -0.5)
Down_CeMbio.yes_CeMbio.no <- Down_CeMbio.yes_CeMbio.no[order(Down_CeMbio.yes_CeMbio.no$padj),]

#################################
Up_CeMbio.yes_OP50.yes <- subset(reslfc_CeMbio.yes_OP50.yes, padj < 0.1 & log2FoldChange > 0.5)
Up_CeMbio.yes_OP50.yes <- Up_CeMbio.yes_OP50.yes[order(Up_CeMbio.yes_OP50.yes$padj),]

Down_CeMbio.yes_OP50.yes <- subset(reslfc_CeMbio.yes_OP50.yes, padj < 0.1 & log2FoldChange > -0.5)
Down_CeMbio.yes_OP50.yes <- Down_CeMbio.yes_OP50.yes[order(Down_CeMbio.yes_OP50.yes$padj),]

#################################
Up_CeMbio.no_OP50.no <- subset(reslfc_CeMbio.no_OP50.no, padj < 0.1 & log2FoldChange > 0.5)
Up_CeMbio.no_OP50.no <- Up_CeMbio.no_OP50.no[order(Up_CeMbio.no_OP50.no$padj),]

Down_CeMbio.no_OP50.no <- subset(reslfc_CeMbio.no_OP50.no, padj < 0.1 & log2FoldChange > -0.5)
Down_CeMbio.no_OP50.no <- Down_CeMbio.no_OP50.no[order(Down_CeMbio.no_OP50.no$padj),]

#################################
Up_OP50.yes_OP50.no <- subset(reslfc_OP50.yes_OP50.no, padj < 0.1 & log2FoldChange > 0.5)
Up_OP50.yes_OP50.no <- Up_OP50.yes_OP50.no[order(Up_OP50.yes_OP50.no$padj),]

Down_OP50.yes_OP50.no <- subset(reslfc_OP50.yes_OP50.no, padj < 0.1 & log2FoldChange > -0.5)
Down_OP50.yes_OP50.no <- Down_OP50.yes_OP50.no[order(Down_OP50.yes_OP50.no$padj),]
```


Save these lists to output files
```{r}
Up_CeMbio.yes_CeMbio.no
Down_CeMbio.yes_CeMbio.no
Up_CeMbio.yes_OP50.yes
Down_CeMbio.yes_OP50.yes
Up_CeMbio.no_OP50.no
Down_CeMbio.no_OP50.no
Up_OP50.yes_OP50.no
Down_OP50.yes_OP50.no

write(rownames(Up_CeMbio.yes_CeMbio.no), file = "03_output/intestine/Up_CeMbio.yes_CeMbio.no.txt", sep = "\n")
write(rownames(Down_CeMbio.yes_CeMbio.no), file = "03_output/intestine/Down_CeMbio.yes_CeMbio.no.txt", sep = "\n")

write(rownames(Up_CeMbio.yes_OP50.yes), file = "03_output/intestine/Up_CeMbio.yes_OP50.yes.txt", sep = "\n")
write(rownames(Down_CeMbio.yes_OP50.yes), file = "03_output/intestine/Down_CeMbio.yes_OP50.yes.txt", sep = "\n")

write(rownames(Up_CeMbio.no_OP50.no), file = "03_output/intestine/Up_CeMbio.no_OP50.no.txt", sep = "\n")
write(rownames(Down_CeMbio.no_OP50.no), file = "03_output/intestine/Down_CeMbio.no_OP50.no.txt", sep = "\n")

write(rownames(Up_OP50.yes_OP50.no), file = "03_output/intestine/Up_OP50.yes_OP50.no.txt", sep = "\n")
write(rownames(Down_OP50.yes_OP50.no), file = "03_output/intestine/Down_OP50.yes_OP50.no.txt", sep = "\n")
```


make hierarchically clustered heatmaps of all changing genes 
```{r}
# get differentially expressed genes
res_CeMbio.yes_CeMbio.no
res_CeMbio.yes_OP50.yes
res_CeMbio.no_OP50.no
res_OP50.yes_OP50.no


#Subset each results table for just the differentially expressed genes:
sign_CeMbio.yes_CeMbio.no <- subset(res_CeMbio.yes_CeMbio.no, padj < 0.05)
dim(subset(res_CeMbio.yes_CeMbio.no, padj < 0.1))

sign_CeMbio.yes_OP50.yes <- subset(res_CeMbio.yes_OP50.yes, padj < 0.05)
dim(subset(res_CeMbio.yes_OP50.yes, padj < 0.1))

sign_CeMbio.no_OP50.no <- subset(res_CeMbio.no_OP50.no, padj < 0.05)
dim(subset(res_CeMbio.no_OP50.no, padj < 0.1))

sign_OP50.yes_OP50.no <- subset(res_OP50.yes_OP50.no, padj < 0.05)
dim(subset(res_OP50.yes_OP50.no, padj < 0.1))


#Determine how many genes were captured and merge them:
changing_genes <- rbind(sign_CeMbio.yes_CeMbio.no, sign_CeMbio.yes_OP50.yes, sign_CeMbio.no_OP50.no, sign_OP50.yes_OP50.no)

changing_exposure <- rbind(sign_CeMbio.yes_CeMbio.no, sign_OP50.yes_OP50.no)

changing_diet <- rbind(sign_CeMbio.no_OP50.no, sign_CeMbio.yes_OP50.yes)


dim(changing_genes)
length(unique(rownames(changing_genes)))

# Get all r-stabilized log values of all the data:
rdl_all <- assay(rlog(dds, blind=FALSE))

#Subset just the changing genes:
changing_all_genes <- subset(rdl_all, rownames(rdl_all) %in% rownames(changing_genes))

changing_exposure_genes <- subset(rdl_all, rownames(rdl_all) %in% rownames(changing_exposure))

changing_diet_genes <- subset(rdl_all, rownames(rdl_all) %in% rownames(changing_diet))


# Make sure it is in matrix form:
class(changing_all_genes)
class(changing_diet_genes)
class(changing_exposure_genes)
```
```{r}
# save plot as a .png
png("03_output/intestine/Heatmap - all genes.png")
# Reset parameters
par(mfrow=c(1,1))
# Draw a heat map
# Scale by row, listed below as scale = "row". This is really important. It sets the mean of every row to 0 and the standard deviation of every row to 1:
pheatmap(changing_all_genes, 
         scale="row", 
         color = colorRampPalette(c("blue", "white", "red"), space = "Lab")(100),
         cluster_rows=FALSE, 
         cluster_cols=TRUE, 
         clustering_distance_rows = "euclidean", 
         clustering_method = "complete",
         #Rowv = NA,
         #Colv = NA,
         show_rownames = FALSE)

dev.off()
```
```{r}
# save plot as a .png
png("03_output/intestine/Heatmap - exposure genes.png")
# Reset parameters
par(mfrow=c(1,1))
# Draw a heat map
# Scale by row, listed below as scale = "row". This is really important. It sets the mean of every row to 0 and the standard deviation of every row to 1:
pheatmap(changing_exposure_genes, 
         scale="row", 
         color = colorRampPalette(c("blue", "white", "red"), space = "Lab")(100),
         cluster_rows=FALSE, 
         cluster_cols=TRUE, 
         clustering_distance_rows = "euclidean", 
         clustering_method = "complete",
         #Rowv = NA,
         #Colv = NA,
         show_rownames = FALSE)

dev.off()
```
```{r}
# save plot as a .png
png("03_output/intestine/Heatmap - diet genes.png")
# Reset parameters
par(mfrow=c(1,1))
# Draw a heat map
# Scale by row, listed below as scale = "row". This is really important. It sets the mean of every row to 0 and the standard deviation of every row to 1:
pheatmap(changing_diet_genes, 
         scale="row", 
         color = colorRampPalette(c("blue", "white", "red"), space = "Lab")(100),
         cluster_rows=FALSE, 
         cluster_cols=TRUE, 
         clustering_distance_rows = "euclidean", 
         clustering_method = "complete",
         #Rowv = NA,
         #Colv = NA,
         show_rownames = FALSE)

dev.off()
```


############################ GO Analysis 

Install Libraries
```{r}
#install.packages("tidyverse")
#if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
#BiocManager::install(version = "3.14")
#BiocManager::install("InterMineR")
```

Load Libraries
```{r, warning=FALSE, message=FALSE}
library(tidyverse)
library(InterMineR)
library(ComplexHeatmap)
library(biomaRt)
```

```{r}
# need to add a column of the WbGeneIDs 
WBgeneID <- rownames(changing_all_genes)
changing_all_genes <- cbind(changing_all_genes, WBgeneID)
changing_all_genes <- as.data.frame(changing_all_genes)

WBgeneID <- rownames(changing_diet_genes)
changing_diet_genes <- cbind(changing_diet_genes, WBgeneID)
changing_diet_genes <- as.data.frame(changing_diet_genes)

WBgeneID <- rownames(changing_exposure_genes)
changing_exposure_genes <- cbind(changing_exposure_genes, WBgeneID)
changing_exposure_genes <- as.data.frame(changing_exposure_genes)

####################
WBgeneID <- rownames(sign_CeMbio.yes_CeMbio.no)
sign_CeMbio.yes_CeMbio.no <- cbind(sign_CeMbio.yes_CeMbio.no, WBgeneID)
sign_CeMbio.yes_CeMbio.no <- as.data.frame(sign_CeMbio.yes_CeMbio.no)

WBgeneID <- rownames(sign_CeMbio.yes_OP50.yes)
sign_CeMbio.yes_OP50.yes <- cbind(sign_CeMbio.yes_OP50.yes, WBgeneID)
sign_CeMbio.yes_OP50.yes <- as.data.frame(sign_CeMbio.yes_OP50.yes)

WBgeneID <- rownames(sign_CeMbio.no_OP50.no)
sign_CeMbio.no_OP50.no <- cbind(sign_CeMbio.no_OP50.no, WBgeneID)
sign_CeMbio.no_OP50.no <- as.data.frame(sign_CeMbio.no_OP50.no)

WBgeneID <- rownames(sign_OP50.yes_OP50.no)
sign_OP50.yes_OP50.no <- cbind(sign_OP50.yes_OP50.no, WBgeneID)
sign_OP50.yes_OP50.no <- as.data.frame(sign_OP50.yes_OP50.no)

####################
# make these too?
Up_CeMbio.yes_CeMbio.no
Down_CeMbio.yes_CeMbio.no
Up_CeMbio.yes_OP50.yes
Down_CeMbio.yes_OP50.yes
Up_CeMbio.no_OP50.no
Down_CeMbio.no_OP50.no
Up_OP50.yes_OP50.no
Down_OP50.yes_OP50.no

######################### also do these, need for later on
reslfc_CeMbio.yes_CeMbio.no
reslfc_CeMbio.yes_OP50.yes
reslfc_CeMbio.no_OP50.no
reslfc_OP50.yes_OP50.no 

WBgeneID <- rownames(reslfc_CeMbio.yes_CeMbio.no)
reslfc_CeMbio.yes_CeMbio.no_data <- cbind(reslfc_CeMbio.yes_CeMbio.no, WBgeneID)
reslfc_CeMbio.yes_CeMbio.no_data <- as.data.frame(reslfc_CeMbio.yes_CeMbio.no_data)

WBgeneID <- rownames(reslfc_CeMbio.yes_OP50.yes)
reslfc_CeMbio.yes_OP50.yes_data <- cbind(reslfc_CeMbio.yes_OP50.yes, WBgeneID)
reslfc_CeMbio.yes_OP50.yes_data <- as.data.frame(reslfc_CeMbio.yes_OP50.yes_data)

WBgeneID <- rownames(reslfc_CeMbio.no_OP50.no)
reslfc_CeMbio.no_OP50.no_data <- cbind(reslfc_CeMbio.no_OP50.no, WBgeneID)
reslfc_CeMbio.no_OP50.no_data <- as.data.frame(reslfc_CeMbio.no_OP50.no_data)

WBgeneID <- rownames(reslfc_OP50.yes_OP50.no)
reslfc_OP50.yes_OP50.no_data <- cbind(reslfc_OP50.yes_OP50.no, WBgeneID)
reslfc_OP50.yes_OP50.no_data <- as.data.frame(reslfc_OP50.yes_OP50.no_data)
```

Get filternames from https://parasite.wormbase.org/biomart/martview
```{r}
paramart <- biomaRt::useMart("parasite_mart", dataset = "wbps_gene", host = "https://parasite.wormbase.org", port = 443)

reslfc_OP50.yes_OP50.no_data <- biomaRt::getBM(
  mart = paramart,
  filter = c("wbps_gene_id"),
  value = reslfc_OP50.yes_OP50.no_data$WBgeneID,
  attributes = c("wbps_gene_id","wormbase_gseq", "wikigene_name")
) %>% right_join(reslfc_OP50.yes_OP50.no_data, by = c("wbps_gene_id" = "WBgeneID"))

reslfc_OP50.yes_OP50.no_data <- reslfc_OP50.yes_OP50.no_data %>% rename(WBgeneID = "wbps_gene_id")

write_csv(reslfc_OP50.yes_OP50.no_data, file = "03_output/intestine/reslfc_OP50.yes_OP50.no_data_sets.csv")
```


Perform gene ontology
```{r}
#install.packages("reticulate")
library(reticulate)
```
```{python}
import pandas as pd
import numpy as np
import tissue_enrichment_analysis as ea
import matplotlib.pyplot as plt
import seaborn as sns
import csv
```
```{python}
tissue = ea.fetch_dictionary('tissue')
phenotype = ea.fetch_dictionary('phenotype')
go = ea.fetch_dictionary('go')
```
```{python}
changing_all_genes=pd.read_csv('03_output/intestine/changing_all_genes_sets.csv')
changing_diet_genes=pd.read_csv('03_output/intestine/changing_diet_genes_sets.csv')
changing_exposure_genes=pd.read_csv('03_output/intestine/changing_exposure_genes_sets.csv')

###############################
sign_CeMbio_yes_CeMbio_no=pd.read_csv('03_output/intestine/sign_CeMbio.yes_CeMbio.no_sets.csv')
sign_CeMbio_yes_OP50_yes=pd.read_csv('03_output/intestine/sign_CeMbio.yes_OP50.yes_sets.csv')
sign_CeMbio_no_OP50_no=pd.read_csv('03_output/intestine/sign_CeMbio.no_OP50.no_sets.csv')
sign_OP50_yes_OP50_no=pd.read_csv('03_output/intestine/sign_OP50.yes_OP50.no_sets.csv')
```
```{python}
one = sign_OP50_yes_OP50_no['WBgeneID']
two = sign_CeMbio_yes_CeMbio_no['WBgeneID']
PA14_background_set = pd.DataFrame(pd.concat([one, two], ignore_index=True))
PA14_background_set.to_csv('03_output/intestine/PA14_background_set.csv', index=False)


three = sign_CeMbio_no_OP50_no['WBgeneID']
four = sign_CeMbio_yes_OP50_yes['WBgeneID']
Diet_background_set = pd.DataFrame({'WBgeneID': pd.concat([three, four], ignore_index=True)})
Diet_background_set.to_csv('03_output/intestine/Diet_background_set.csv', index=False)
```
```{python}
frames = {'tissue': tissue, 'phenotype': phenotype, 'go': go}
###############################
PA14_background_set
Diet_background_set

changing_all_genes
changing_diet_genes
changing_exposure_genes

sign_CeMbio_yes_CeMbio_no
sign_CeMbio_yes_OP50_yes
sign_CeMbio_no_OP50_no
sign_OP50_yes_OP50_no
###############################
# original code
result = {}
for analysis, dictionary in frames.items():
  result[analysis] = ea.enrichment_analysis(changing_exposure_genes.WBgeneID, dictionary, show=True, alpha=10**-1)


# compare to the generated background set 
#result = {}
#for analysis, dictionary in frames.items():
#  result[analysis] = ea.enrichment_analysis(sign_CeMbio_yes_CeMbio_no.WBgeneID, dictionary[dictionary.wbid.isin(PA14_background_set.WBgeneID)], show=True, alpha=10**-1)
  
  
# make the figure in the paper:
fig, ax = plt.subplots(nrows=3, figsize=(8, 10), facecolor='white')
i= 0

# go through the results hash:
for t, r in result.items():
    # calculate the negative log of the Q-values and store
    r['logQ'] = -r['Q value'].apply(np.log10)
    # remove np.infinites with np.nan
    r.logQ.replace([np.inf], np.nan, inplace=True)
    # remove np.nan with 70 (after 10**-64, the hypergeometric function crashes and returns 0)
    r.logQ.replace(np.nan, 70, inplace=True)
    
    # call the plotting function in the Enrichment Suite to plot the results
    ea.plot_enrichment_results(r, title=t, analysis=t, ax=ax[i], y='logQ', n_bars=10, color=(0.5, 0.5, 0.5))

    # prettify axes
    ax[i].set_ylabel(t)
    if i != 2:
        ax[i].set_xlabel('')
    else:
        ax[i].set_xlabel(r'$-\log_{10}{Q}$')
    i += 1

# save figure
plt.savefig('03_output/intestine/changing_exposure_genes_Enrichment_Results.svg', bbox_inches='tight')  

# save the results from GO to a csv
results = result["go"]
results.to_csv('03_output/intestine/Results-changing_exposure_genes-comparison.csv', index=False)
```

Read in GO results from the comparisons within groups
```{r}
sign_CeMbio_yes_CeMbio_no
sign_CeMbio_yes_OP50_yes
sign_CeMbio_no_OP50_no
sign_OP50_yes_OP50_no


# GO results
sign_CeMbio_yes_CeMbio_no_comparison <- read.csv2("03_output/intestine/Results-sign_CeMbio_yes_CeMbio_no-comparison.csv", header = TRUE, sep = ",") %>% arrange(logQ)

sign_CeMbio_yes_OP50_yes_comparison <- read.csv2("03_output/intestine/Results-sign_CeMbio_yes_OP50_yes-comparison.csv", header = TRUE, sep = ",") %>% arrange(logQ)

sign_CeMbio_no_OP50_no_comparison <- read.csv2("03_output/intestine/Results-sign_CeMbio_no_OP50_no-comparison.csv", header = TRUE, sep = ",") %>% arrange(logQ)

sign_OP50_yes_OP50_no_comparison <- read.csv2("03_output/intestine/Results-sign_OP50_yes_OP50_no-comparison.csv", header = TRUE, sep = ",") %>% arrange(logQ)
```

### this still needs tweaking ###
Plot the GO comparisons
```{r}
library(ggplot2)

sign_CeMbio_yes_CeMbio_no_comparison
sign_CeMbio_yes_OP50_yes_comparison
sign_CeMbio_no_OP50_no_comparison
sign_OP50_yes_OP50_no_comparison

# save plot as a .png
png("03_output/intestine/GO_Usign_CeMbio_yes_CeMbio_no_comparison.png")
sign_CeMbio_yes_CeMbio_no_comparison %>% top_n(10, logQ) %>%
ggplot(aes(x = reorder(Term, desc(logQ)), y = logQ)) +
  geom_bar(stat = "identity", fill = "#808080") +
  coord_flip() +
  labs(title = "GO Analysis", x = "Term", y = expression("-log"[10]*"{Q}")) + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))

dev.off()
```

creat an upset plot
```{r}
#install.packages("UpSetR")
#BiocManager::install("ComplexHeatmap")
library(UpSetR)
library(ComplexHeatmap)

changing_all_genes
changing_diet_genes
changing_exposure_genes

changing_all_genes <- read.csv2("03_output/intestine/changing_all_genes_sets.csv", header = TRUE, sep = ",")
changing_diet_genes <- read.csv2("03_output/intestine/changing_diet_genes_sets.csv", header = TRUE, sep = ",")
changing_exposure_genes <- read.csv2("03_output/intestine/changing_exposure_genes_sets.csv", header = TRUE, sep = ",")
########################

sign_CeMbio.yes_CeMbio.no
sign_CeMbio.yes_OP50.yes
sign_CeMbio.no_OP50.no
sign_OP50.yes_OP50.no

sign_CeMbio.yes_CeMbio.no <- read.csv2("03_output/intestine/sign_CeMbio.yes_CeMbio.no_sets.csv", header = TRUE, sep = ",")

sign_CeMbio.yes_OP50.yes <- read.csv2("03_output/intestine/sign_CeMbio.yes_OP50.yes_sets.csv", header = TRUE, sep = ",")

sign_CeMbio.no_OP50.no <- read.csv2("03_output/intestine/sign_CeMbio.no_OP50.no_sets.csv", header = TRUE, sep = ",")

sign_OP50.yes_OP50.no <- read.csv2("03_output/intestine/sign_OP50.yes_OP50.no_sets.csv", header = TRUE, sep = ",")
```
```{r}
lt <- list(changing_all_genes = changing_all_genes$WBgeneID, 
           changing_diet_genes = changing_diet_genes$WBgeneID, 
           changing_exposure_genes = changing_exposure_genes$WBgeneID)

lt <- list(CeMbio.yes_CeMbio.no = sign_CeMbio.yes_CeMbio.no$WBgeneID, 
           CeMbio.yes_OP50.yes = sign_CeMbio.yes_OP50.yes$WBgeneID,
           CeMbio.no_OP50.no = sign_CeMbio.no_OP50.no$WBgeneID,
           OP50.yes_OP50.no = sign_OP50.yes_OP50.no$WBgeneID)

str(lt)
list_to_matrix(lt)

m1 = make_comb_mat(lt)
#m1 = make_comb_mat(lt2)
set_name(m1)
comb_name(m1)
set_size(m1)
comb_size(m1)
comb_degree(m1)
t(m1)

ss = set_size(m1)
cs = comb_size(m1)

plot_t <- UpSet(m1, 
      set_order = order(ss),
      comb_order = order(comb_degree(m1), -cs),
      comb_col = c("blue", "lightblue", "black")[comb_degree(m1)],
      left_annotation = upset_left_annotation(m1))

# save plot as a .png
png("03_output/intestine/Upset_plot_group_comparisons.png")
plot_t

dev.off()
```

-----------------------------------------------------

MA plot
```{r}
#install.packages("ggpubr")
library(ggpubr)
#install.packages("readxl")
library(readxl)
#install.packages("ggrepel")
library(ggrepel)
```
```{r}
# read in the data and convert to a dataframe
reslfc_CeMbio.yes_CeMbio.no_data_sets <- read_delim("03_output/intestine/reslfc_CeMbio.yes_CeMbio.no_data_sets.csv", delim = ",", col_names = TRUE)
reslfc_CeMbio.yes_CeMbio.no_data_sets <- as.data.frame(reslfc_CeMbio.yes_CeMbio.no_data_sets)

reslfc_CeMbio.yes_OP50.yes_data_sets <- read_delim("03_output/intestine/reslfc_CeMbio.yes_OP50.yes_data_sets.csv", delim = ",", col_names = TRUE)
reslfc_CeMbio.yes_OP50.yes_data_sets <- as.data.frame(reslfc_CeMbio.yes_OP50.yes_data_sets)

reslfc_CeMbio.no_OP50.no_data_sets <- read_delim("03_output/intestine/reslfc_CeMbio.no_OP50.no_data_sets.csv", delim = ",", col_names = TRUE)
reslfc_CeMbio.no_OP50.no_data_sets <- as.data.frame(reslfc_CeMbio.no_OP50.no_data_sets)

reslfc_OP50.yes_OP50.no_data_sets <- read_delim("03_output/intestine/reslfc_OP50.yes_OP50.no_data_sets.csv", delim = ",", col_names = TRUE)
reslfc_OP50.yes_OP50.no_data_sets <- as.data.frame(reslfc_OP50.yes_OP50.no_data_sets)
```
```{r}
clec_IDs <- read_excel("03_output/intestine/clec_WBgeneIDs.xlsx", col_names = TRUE, sheet = 2)
class(clec_IDs)
clec_IDs <- as.data.frame(clec_IDs)
head(clec_IDs)
```
```{r}
# Save the plot
png("03_output/intestine/MA plot - OP50.yes_OP50.no_clecs.png")
# Reset plotting parameters to their default values
par(mfrow=c(1,1))
# plot the data
ggmaplot(reslfc_OP50.yes_OP50.no_data_sets, 
         main = "OP50.yes vs OP50.no",
   fdr = 0.05, fc = 1.5, size = 0.4,
   palette = c("#B31B21", "#1465AC", "darkgray"),
   genenames = NULL,
   legend = "top",
   font.label = c("bold", 11),
   font.legend = "bold",
   font.main = "bold",
   ggtheme = ggplot2::theme_minimal(),
   top = 0) + geom_label_repel(data = subset(reslfc_OP50.yes_OP50.no_data_sets, wikigene_name %in% clec_IDs$Gene_name), #max.overlaps = getOption("ggrepel.max.overlaps", default = 10),
               mapping = aes(label = as.vector(wikigene_name),
                             x = log2(baseMean),
                             y = log2FoldChange),
               box.padding = unit(0.35, "lines"),
               point.padding = unit(0.3, "lines"),
               force = 1,
               fontface = "bold",
               size = 10/3)
dev.off()
```
```{r}
reslfc_CeMbio.yes_CeMbio.no_data_sets
reslfc_CeMbio.yes_OP50.yes_data_sets
reslfc_CeMbio.no_OP50.no_data_sets
reslfc_OP50.yes_OP50.no_data_sets

# Save the plot
png("03_output/intestine/MA plot - CeMbio.yes_CeMbio.no_top20.png")
# Reset plotting parameters to their default values
par(mfrow=c(1,1))
# plot the data
ggmaplot(reslfc_CeMbio.yes_CeMbio.no_data_sets, 
         main = "CeMbio.yes vs CeMbio.no",
   fdr = 0.05, fc = 1.5, size = 0.4,
   palette = c("#B31B21", "#1465AC", "darkgray"),
   genenames = reslfc_CeMbio.yes_CeMbio.no_data_sets$wikigene_name,
   top = 20,
   font.label = c("bold", 11),
   font.legend = "bold",
   font.main = "bold",
   ggtheme = ggplot2::theme_minimal()) 

dev.off()
```

Volcano plots
```{r}
#Identify significantly changing genes
significantLFC.diet.int <- rbind(subset(resLFC_diet.intestine, padj < 0.1 & log2FoldChange > 0.5), subset(resLFC_diet.intestine, padj < 0.1 & log2FoldChange < -0.5))

significant_points_to_plot.diet.int <-resLFC_diet.intestine[which(rownames(resLFC_diet.intestine) %in% rownames(significantLFC.diet.int)),] 


significantLFC.exposure.int <- rbind(subset(resLFC_exposure.intestine, padj < 0.1 & log2FoldChange > 0.5), subset(resLFC_exposure.intestine, padj < 0.1 & log2FoldChange < -0.5))

significant_points_to_plot.exposure.int <-resLFC_exposure.intestine[which(rownames(resLFC_exposure.intestine) %in% rownames(significantLFC.exposure.int)),] 

# We will set the top limit of the plot as 100. Need to find all the points that exceed that measure
maxedout.diet.int <- subset(resLFC_diet.intestine, padj < 10e-100)
maxedout.exposure.int <- subset(resLFC_exposure.intestine, padj < 10e-100)


png("03_output/volcano plot - diet (int).png")
# Reset parameters
par(mfrow=c(1,1))
# Draw the plot
ggplot(resLFC_exposure.intestine$log2FoldChange, -log10(resLFC_exposure.intestine$padj),
     main="Volcano plot - exposure (int)", xlab="Effect size: log2(fold-change)", ylab="-log10(adjusted p-value)", pch=20, cex = 0.75, ylim = c(0, 20), xlim = c(-10,10), col = "#00000030")
# Add points and labels
points(maxedout.exposure.int$log2FoldChange, rep(102, length(maxedout.exposure.int$log2FoldChange)), 
       pch=20, cex = 0.75, ylim = c(0, 100), col = "green")
points(significant_points_to_plot.exposure.int$log2FoldChange, -log10(significant_points_to_plot.exposure.int$padj),
       pch=20, cex = 0.75, ylim = c(0, 100), col = "red")
labels = rownames(resLFC_exposure.intestine)[rownames(resLFC_exposure.intestine) %in% clec_WBgeneIDs]
# Add lines
abline(v=0, col = "blue")
abline(v=0.5, col = "blue", lty = "dashed")
abline(v=-0.5, col = "blue", lty = "dashed")
abline(h=-log10(0.1), col = "blue", lty = "dashed")

dev.off()
```


